nexus-sdk/src/main/java/io/nexusrpc/OperationHandler.java

```diff
@@ -0,0 +1,228 @@
+ */
+public interface OperationHandler<T, R> {
+  /**
+   * Create an operation handler for a synchronous operation backed by the given function. This
+   * function will be called for each operation invocation. All other calls to the operation are not
+   * supported.
+   */
+  static <T, R> OperationHandler<T, R> sync(BiFunction<StartContext, @Nullable T, R> func) {
+    return new OperationHandler<T, R>() {
+      @Override
+      public StartResult<R> start(StartContext ctx, @Nullable T param) {
+        return StartResult.sync(func.apply(ctx, param));
+      }
+
+      @Override
+      public R fetchResult(FetchResultContext ctx) {
+        throw new UnsupportedOperationException("Not supported on sync operation");
+      }
+
+      @Override
+      public OperationInfo fetchInfo(FetchInfoContext ctx) {
+        throw new UnsupportedOperationException("Not supported on sync operation");
+      }
+
+      @Override
+      public void cancel(CancelContext ctx) {
+        throw new UnsupportedOperationException("Not supported on sync operation");
+      }
+    };
+  }
+
+  /**
+   * Handle the start of an operation.
+   *
+   * <p>The operation can be synchronous or asynchronous. Synchronous operations can return the
+   * result via {@link StartResult#sync}. Asynchronous operations can return the started operation
+   * ID via {@link StartResult#async}.
+   *
+   * @param ctx Context containing details about the call.
+   * @param param Parameter for the operation. This may be null if the parameter was not given.
+   * @return The start result which is a synchronous value or an operation ID representing an
+   *     asynchronously running operation.
+   * @throws OperationUnsuccessfulException If thrown, can have failure details and state such as
+   *     saying the operation was cancelled.
+   * @throws RuntimeException Any other exception fails the operation start as an operation failure.
+   */
+  StartResult<R> start(StartContext ctx, @Nullable T param) throws OperationUnsuccessfulException;
+
+  /**
+   * Fetch the result for an asynchronously started operation.
+   *
+   * @param ctx Context containing details about the call including the operation ID. The context
+   *     also contains a timeout which affects how implementers should implement this function. See
+   *     {@link FetchResultContext#getTimeout()} to see how to react to this value.
+   * @return The resulting value upon success.
+   * @throws OperationNotFoundException Operation ID provided is not a known ID.
+   * @throws OperationStillRunningException Operation is still running beyond the given timeout.
+   * @throws OperationUnsuccessfulException Operation failed. If thrown, can have failure details
+   *     and state such as saying the operation was cancelled.
+   * @throws RuntimeException Any other exception is considered a failure to fetch the result.
+   */
+  R fetchResult(FetchResultContext ctx)
```

Comment:
Did we make an explicit decision to use checked exceptions? The Java SDK almost never uses checked exception. I am fine with using them here though.

Reply:
I don't necessarily think we should use checked exceptions for most things, but for these use cases I think we should. Granted only we call these handler methods, but I expect these same exceptions are used client side (well, at least `OperationNotFoundException` and `OperationUnsuccessfulException`). I still expect unchecked exceptions to occur on network things, but I think these represent something we want to force the caller to catch or declare. I am usually against checked exceptions if I think the caller is just going to catch and rethrow as a `RuntimeException` all the time, but I think these should get more explicit behavior.

Having said all of that, I am ok changing these to unchecked if we want (though I'd keep the throws/docs here).